### PR TITLE
centrifuge: Anemoy pool currency migration

### PR DESCRIPTION
**Pull request by NunoAlexandre**
# Description

We want to migrate the `currency` of the `Anemoy` pool from `LpEthUSDC` to (Polkadot-native) `USDC` on Centrifuge.

- Anemoy has pool id 4,139,607,887 (the only pool currently registered on Centrifuge)
- We will migrate the currency of that pool from `ForeignAsset(100,001)` (`LpEthUSDC`) to `ForeignAsset(6)`  (`Circle’s USDC native on Polkadot`)
- Both currencies have 6 decimals so no extra considerations in that regard

## To Do

- [x] You can just add a sanity check to the migration that the pool value is indeed 0 and there are no locked orders
  - [x] For all tranches within the Anemoy pool, verify that are no entries on pallet_investments' `InvestOrders` , `RedeemOrders` , `ActiveInvestOrders` and `ActiveRedeemOrders`.
  - [x] Check that pool value is `0`

### CODE REVIEW

File: libs/types/src/tokens.rs [hunk 1]
```diff
@@ -187,8 +187,8 @@ where
 )]
 #[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
 pub struct TrancheCurrency {
-	pub(crate) pool_id: PoolId,
-	pub(crate) tranche_id: TrancheId,
+	pub pool_id: PoolId,
```
Comment (NunoAlexandre) [inline libs/types/src/tokens.rs:190]: any objections to open these fields visibility? by having them `pub(crate)` I would have to import he `TrancheCurrencyT` trait to call `generate` to build an instance of this type which is quite an overkill imo. let me know if I am not overseeing any requirement 👍

File: runtime/centrifuge/src/migrations.rs [hunk 2]
```diff
@@ -9,5 +9,92 @@
...
+
+			log::info!("anemoy_pool::Migration: post_upgrade succeeded");
+			Ok(())
+		}
+	}
+
+	// todo(nuno): also check that pool value is 0 and check also that Investments::InvestOrders and
+	// Investments::RedeemOrder have no entries from Anemoy; the latter ones seem tricky at first
+	// sight since they are double maps first keyed by an AccountId, meaning we need to transverse
+	// that first which is more costly.
```
Comment (wischli) [inline runtime/centrifuge/src/migrations.rs:93]: Should suffice to check `StorageDoubleMap::<T>::iter_keys().count().is_zero()`.

Comment (NunoAlexandre) [reply runtime/centrifuge/src/migrations.rs:93]: Can you expand on that? Because we need to iter keys on the account (first storage key) and then filter by the investment id (pool and tranche); if we just do iter_keys we are of taking any other pool into account.

If the idea is to use the assumption that anemoy is the only pool, then it makes sense. But it sort of defeats or weakens the point of this sanity check in the first place (tho, sure, it's less likely there will be another pool being added than a new order for anemoy).

Wtdyt? If everyone is fine with that I'm not blocking the approach

Comment (NunoAlexandre) [reply runtime/centrifuge/src/migrations.rs:93]: I'll just filter the iter_keys involving anemoy to make this 100% robust.

**Comment by NunoAlexandre** [comment]
Implemented the sanity checks for orders under the Anemoy pool and tested again with `try-runtime`. Looking good:

```
2023-09-23 14:15:57.471  INFO main centrifuge_runtime::migrations::anemoy_pool: anemoy_pool::Migration: currency set to USDC ✓
2023-09-23 14:15:57.471  INFO main centrifuge_runtime::migrations::anemoy_pool: anemoy_pool::Migration: post_upgrade succeeded ✓
```

Still have to implement the check for the pool value being 0.

File: runtime/centrifuge/src/migrations.rs [hunk 3]
```diff
@@ -9,5 +9,126 @@
...
+			pallet_investments::ActiveInvestOrders::<Runtime>::iter_keys()
+				.filter(|investment| investment.pool_id == ANEMOY_POOL_ID)
+				.count() == 0 && pallet_investments::ActiveInvestOrders::<Runtime>::iter_keys()
+				.filter(|investment| investment.pool_id == ANEMOY_POOL_ID)
+				.count() == 0 && pallet_investments::InvestOrders::<Runtime>::iter_keys()
+				.filter(|(_, investment)| investment.pool_id == ANEMOY_POOL_ID)
+				.count() == 0 && pallet_investments::RedeemOrders::<Runtime>::iter_keys()
+				.filter(|(_, investment)| investment.pool_id == ANEMOY_POOL_ID)
+				.count() == 0;
+
```
Comment (mustermeiszer) [inline runtime/centrifuge/src/migrations.rs:118]: That looks soo clean! nice!

**Review by mustermeiszer — DISMISSED** [review]
Looks good!

File: runtime/centrifuge/src/migrations.rs (hunk 3 above)
Comment (NunoAlexandre) [reply runtime/centrifuge/src/migrations.rs]: thanks 🙏 just pushed the check for pool value == 0 and realised that we were checking `ActiveInvestOrders` twice instead of checking `ActiveRedeemOrders` :) fixed

**Comment by NunoAlexandre** [comment]
This PR is now feature-complete. Just testing it one last time 👍

File: pallets/investments/src/lib.rs [hunk 4]
```diff
@@ -224,7 +224,7 @@ pub mod pallet {
 		StorageMap<_, Blake2_128Concat, T::InvestmentId, OrderId, ValueQuery>;
 
 	#[pallet::storage]
-	pub(crate) type InvestOrders<T: Config> = StorageDoubleMap<
+	pub type InvestOrders<T: Config> = StorageDoubleMap<
```
Comment (wischli) [inline pallets/investments/src/lib.rs:227]: This enables writing these storages from outside which is not desired. We should at least add a note to revert this change after removing the migration. 

Since we only read, it appears to me that it would be beneficial on using existing /adding `getters`.

File: runtime/centrifuge/src/migrations.rs [hunk 5]
```diff
@@ -9,5 +9,127 @@
...
+				log::info!("anemoy_pool::Migration: currency set to USDC ✓");
+			});
+
+			weight.saturating_add(
+				<Runtime as frame_system::Config>::DbWeight::get().reads(
+					pallet_pool_system::Pool::<Runtime>::iter_keys()
+						.count()
+						.saturating_mul(2) as u64,
+				),
+			)
```
Comment (wischli) [inline runtime/centrifuge/src/migrations.rs:83]: Why are we adding two reads multiplied with the number of pool keys? Shouldn't it just add single read and a write?

**Review by wischli — COMMENTED** [review]
The migration LGTM! I would like to raise my concerns about making the `pallet-investment` storages public.

File: runtime/centrifuge/src/migrations.rs (hunk 5 above)
Comment (NunoAlexandre) [reply runtime/centrifuge/src/migrations.rs:83]: fixed. I guess the result of moving coding around and changes to the implementation. Made this add 1 read and 1 write here, for the pool lookup and mutate, respectively. thanks!

File: pallets/investments/src/lib.rs (hunk 4 above)
Comment (NunoAlexandre) [reply pallets/investments/src/lib.rs:227]: Opened an issue to take care of this and other things after the upgrade: rhttps://github.com/centrifuge/centrifuge-chain/issues/1567

We have full control of what code makes it's way to production so it's not that opening up this visibility puts us at danger that ill-intended users or ourselves corrupting those storages. Adding `getters` imo is just adding code that's not needed aside from this migration that's to be executed once and deleted after so I'd prefer to keep this as is.

**Review by wischli — APPROVED** [review]
LGTM! Thanks for adding the post-merge issue!
